fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanValidatorImpl.java

```diff
@@ -0,0 +1,79 @@
+ * under the License.
+ */
+package org.apache.fineract.portfolio.loanaccount.serialization;
+
+import com.google.gson.JsonElement;
+import java.math.BigDecimal;
+import lombok.RequiredArgsConstructor;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.fineract.infrastructure.core.api.JsonCommand;
+import org.apache.fineract.infrastructure.core.exception.InvalidJsonException;
+import org.apache.fineract.infrastructure.core.serialization.FromJsonHelper;
+import org.apache.fineract.infrastructure.core.service.MathUtil;
+import org.apache.fineract.portfolio.common.service.Validator;
+import org.apache.fineract.portfolio.loanaccount.api.LoanApiConstants;
+import org.apache.fineract.portfolio.loanaccount.domain.Loan;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanDisbursementDetails;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanRepository;
+import org.apache.fineract.portfolio.loanaccount.exception.LoanNotFoundException;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public final class LoanValidatorImpl implements LoanValidator {
+
+    private final FromJsonHelper fromApiJsonHelper;
+    private final LoanRepository loanRepository;
+
+    @Override
+    public void validateLoanApprovedAmountModification(JsonCommand command) {
+        String json = command.json();
+        if (StringUtils.isBlank(json)) {
+            throw new InvalidJsonException();
+        }
+
+        Validator.validateOrThrowDomainViolation("loan.approved.amount", baseDataValidator -> {
+            final JsonElement element = this.fromApiJsonHelper.parse(json);
+            final BigDecimal newApprovedAmount = this.fromApiJsonHelper
+                    .extractBigDecimalWithLocaleNamed(LoanApiConstants.amountParameterName, element);
+
+            baseDataValidator.reset().parameter(LoanApiConstants.amountParameterName).value(newApprovedAmount).positiveAmount();
+
+            final Long loanId = command.getLoanId();
+            Loan loan = this.loanRepository.findById(loanId).orElseThrow(() -> new LoanNotFoundException(loanId));
+            BigDecimal approvedPrincipal = loan.getApprovedPrincipal();
+
+            if (MathUtil.isGreaterThan(newApprovedAmount, approvedPrincipal)) {
+                baseDataValidator.reset().parameter(LoanApiConstants.amountParameterName).failWithCode("greater.than.approved.principal");
+            }
+
+            BigDecimal reservedPrincipal = BigDecimal.ZERO;
+            if (!loan.loanProduct().isDisallowExpectedDisbursements()) {
+                reservedPrincipal = loan.getExpectedLoanDisbursementDetails().stream().map(LoanDisbursementDetails::principal)
+                        .reduce(BigDecimal.ZERO, BigDecimal::add);
+            }
+
+            reservedPrincipal = reservedPrincipal.add(loan.getSummary().getTotalPrincipal());
+
+            if (MathUtil.isLessThan(newApprovedAmount, reservedPrincipal)) {
```

Comment:
I am not fond of "reserved principal" definition. Would "total already disbursed and expected disbursement amount is higher than new approved amount"?